ffi/src/scan.rs

```diff
@@ -0,0 +1,392 @@
+///
+/// The iterator must be valid (returned by [kernel_scan_data_init]) and not yet freed by
+/// [kernel_scan_data_free]. The visitor function pointer must be non-null.
+#[no_mangle]
+pub unsafe extern "C" fn kernel_scan_data_next(
+    data: &mut KernelScanDataIterator,
+    engine_context: *mut c_void,
+    engine_visitor: extern "C" fn(
+        engine_context: *mut c_void,
+        engine_data: *mut EngineDataHandle,
+        selection_vector: KernelBoolSlice,
+    ),
+) -> ExternResult<bool> {
+    kernel_scan_data_next_impl(data, engine_context, engine_visitor)
+        .into_extern_result(data.engine_interface.error_allocator())
+}
+fn kernel_scan_data_next_impl(
+    data: &mut KernelScanDataIterator,
+    engine_context: *mut c_void,
+    engine_visitor: extern "C" fn(
+        engine_context: *mut c_void,
+        engine_data: *mut EngineDataHandle,
+        selection_vector: KernelBoolSlice,
+    ),
+) -> DeltaResult<bool> {
+    if let Some((data, sel_vec)) = data.data.next().transpose()? {
+        let bool_slice: KernelBoolSlice = sel_vec.into();
+        let data_handle = BoxHandle::into_handle(EngineDataHandle { data });
+        (engine_visitor)(engine_context, data_handle, bool_slice);
+        // ensure we free the data
+        unsafe { BoxHandle::drop_handle(data_handle) };
+        Ok(true)
+    } else {
+        Ok(false)
+    }
+}
+
+/// # Safety
+///
+/// Caller is responsible for (at most once) passing a valid pointer returned by a call to
+/// [kernel_scan_files_init].
+// we should probably be consistent with drop vs. free on engine side (probably the latter is more
+// intuitive to non-rust code)
+#[no_mangle]
+pub unsafe extern "C" fn kernel_scan_data_free(data: *mut KernelScanDataIterator) {
+    BoxHandle::drop_handle(data);
+}
+
+type CScanCallback = extern "C" fn(
+    engine_context: *mut c_void,
+    path: KernelStringSlice,
+    size: i64,
+    dv_info: *mut CDvInfo,
+    partition_map: *mut CStringMap,
+);
+
+pub struct CDvInfo {
+    dv_info: DvInfo,
+}
+impl BoxHandle for CDvInfo {}
+
+pub struct CStringMap {
+    values: HashMap<String, String>,
+}
+impl BoxHandle for CStringMap {}
+
+#[no_mangle]
+/// allow probing into a CStringMap. If the specified key is in the map, kernel will call
+/// allocate_fn with the value associated with the key and return the value returned from that
+/// function. If the key is not in the map, this will return NULL
+///
+/// # Safety
+///
+/// The engine is responsible for providing a valid [`CStringMap`] pointer and [`KernelStringSlice`]
+pub unsafe extern "C" fn get_from_map(
+    raw_map: *mut CStringMap,
+    key: KernelStringSlice,
+    allocate_fn: AllocateStringFn,
+) -> *mut c_void {
+    asbox!(raw_map as boxed_map => {
+        let string_key = String::try_from_slice(key);
+        match boxed_map.values.get(&string_key) {
+            Some(v) => {
+                let slice: KernelStringSlice = v.as_str().into();
+                allocate_fn(slice)
```

Comment:
My concern with having an `Into<KernelStringSlice> for String` would be tracking the lifetimes and freeing. Like for `&str` we assume that the slice is NOT owned, but for a `String` it would be. so now you have to think about which kind you have and how you need to free it.

I think for now leaving it with just `&str` makes more sense.